### PR TITLE
fern: round10 — model-slices=256 sweep (double slice count)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

The Transolver backbone uses **slice attention** — the `--model-slices` parameter controls how many slices the model partitions the input into. Tay has used 128 slices for all experiments. Doubling to **256 slices** gives the model finer-grained spatial grouping, which could help capture the high-frequency gradients in tau_y/tau_z that dominate the binding gap (current error 3.7-3.9× AB-UPT reference).

Hypothesis: more slices → finer spatial resolution in attention → better wall-shear field capture → improved tau_y/tau_z. Single delta: `--model-slices 256` (double current).

Note: 256 slices adds ~O(slices) attention keys, increasing memory slightly but not dramatically. Expect similar per-epoch runtime.

## Instructions

Run the SOTA reproduce command with `--model-slices 256`. All other hyperparameters unchanged from PR #111 SOTA.

```bash
cd target/
torchrun --standalone --nproc_per_node=8 train.py \
  --agent fern \
  --optimizer lion --lion-beta1 0.9 --lion-beta2 0.99 \
  --lr 5e-5 --weight-decay 5e-4 \
  --no-compile-model \
  --volume-loss-weight 2.0 --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 256 \
  --ema-decay 0.999 --cosine-t-max 50 \
  --gradient-log-every 100 --weight-log-every 100 \
  --no-log-gradient-histograms \
  --wandb-group tay-round10-slices-256 \
  --wandb-name "fern/round10-slices-256"
```

**Single delta vs SOTA PR #111:** `--model-slices 256` (was `128`). Everything else preserved.

If OOM at 256 slices + 65k pts, try `--model-slices 256` with `--train-surface-points 60000 --eval-surface-points 60000 --train-volume-points 60000 --eval-volume-points 60000` and report.

Hard timeout: `SENPAI_TIMEOUT_MINUTES` (do NOT override). Hard epoch cap: `SENPAI_MAX_EPOCHS` (do NOT override).

## Baseline metrics (current SOTA — PR #111 tanjiro EMA=0.999)

| Metric | tay SOTA (PR #111) | AB-UPT ref |
|---|---:|---:|
| `abupt` mean | **11.142** | — |
| `surface_pressure` | 6.209 | 3.82 |
| `wall_shear` | 11.138 | 7.29 |
| `volume_pressure` | 12.548 | 6.08 |
| `tau_x` | 9.436 | 5.35 |
| `tau_y` | 13.525 | 3.65 |
| `tau_z` | 13.992 | 3.63 |

Best val: 9.989 (ep9). W&B run `ab3y4ej7`.

## Results

_(to be filled in after training completes)_

Post: best `val_primary/abupt_axis_mean_rel_l2_pct`, final `test_primary/abupt_axis_mean_rel_l2_pct`, per-component test breakdown, W&B run URL.

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT fern:

## Results

**Hypothesis FAILED — doubling slices to 256 worsens every metric vs the 128-slice PR #111 SOTA.**

W&B run: [`9lc1acwf`](https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/9lc1acwf) (group `tay-round10-slices-256`)

### Test metrics (best-val ckpt reload)

| Metric | This (256 slices) | PR #111 SOTA (128) | AB-UPT ref | Δ vs SOTA |
|---|---:|---:|---:|---:|
| `abupt_axis_mean_rel_l2_pct` | **12.389** | 11.142 | — | **+1.247 (+11.2%)** |
| `surface_pressure_rel_l2_pct` | 7.121 | 6.209 | 3.82 | +0.912 (+14.7%) |
| `wall_shear_rel_l2_pct` | 12.616 | 11.138 | 7.29 | +1.478 (+13.3%) |
| `volume_pressure_rel_l2_pct` | 13.107 | 12.548 | 6.08 | +0.559 (+4.4%) |
| `wall_shear_x_rel_l2_pct` | 10.773 | 9.436 | 5.35 | +1.337 (+14.2%) |
| `wall_shear_y_rel_l2_pct` | 15.237 | 13.525 | 3.65 | +1.712 (+12.7%) |
| `wall_shear_z_rel_l2_pct` | 15.709 | 13.992 | 3.63 | +1.717 (+12.3%) |

Test-primary MAEs: surface_pressure 0.01996, wall_shear 0.11639, volume_pressure 28.795.

### Validation
- Best `val_primary/abupt_axis_mean_rel_l2_pct` = **11.210** at epoch 7 (vs 9.989 at epoch 9 for 128-slice SOTA).
- Best-checkpoint source: `ema`. Selection metric updated 1× before plateau.

### Run details
- Total runtime: **270.99 min** (timeout-bounded; pod env had `train_timeout_minutes=270`).
- Epochs completed: ~30 of 50 planned (`epoch_time_s ≈ 535s` / ~8.92 min/epoch).
- Global step: 16,891.
- Peak VRAM: **~94.4 GB / 97 GB** (98.8 % allocated on rank 0/1/7) — close to OOM but no crash.

### Command actually executed
The PR command had a flag typo `--cosine-t-max 50`. The valid flag is `--lr-cosine-t-max`, so the run launched on the corrected variant — config otherwise matches the PR verbatim:

```bash
cd target/
torchrun --standalone --nproc_per_node=8 train.py \
  --agent fern \
  --optimizer lion --lion-beta1 0.9 --lion-beta2 0.99 \
  --lr 5e-5 --weight-decay 5e-4 \
  --no-compile-model \
  --volume-loss-weight 2.0 --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 256 \
  --ema-decay 0.999 --lr-cosine-t-max 50 \
  --gradient-log-every 100 --weight-log-every 100 \
  --no-log-gradient-histograms \
  --wandb-group tay-round10-slices-256 \
  --wandb-name "fern/round10-slices-256"
```

### What happened
- The hypothesis predicted that finer spatial grouping (more slices) would help capture high-frequency tau_y/tau_z gradients. Instead **tau_y/tau_z got worse by ~12.5 %**, in line with the rest of the metrics. Doubling slices was uniformly harmful, not selectively helpful for the high-frequency axes.
- Best validation peaked **2 epochs earlier** (7 vs 9) at a **worse** value (11.21 vs 9.99) than 128-slice SOTA. Same EMA, same LR schedule, same step budget — slice attention with 256 slices on a 4 L × 512 d × 8 h backbone does not improve on 128.
- Likely cause: with 65,536 surface tokens projected into slice space, doubling slices halves the effective tokens-per-slice, making per-slice features noisier and harder to optimize within the same step budget. The `s` heatmap (slice assignment) is also a softmax over a doubled key space, which dilutes assignment confidence.
- Slice attention memory grows roughly linearly in `--model-slices`, but K/V/intermediate buffers in the slice-attention block pushed peak VRAM to 98.8 %. We dodged OOM at 65,536 pts with no compile, but the buffer is razor thin.

### Suggested follow-ups
- Combine `--model-slices 256` with `--model-heads 16` (or `--model-hidden-dim 768`) so each slice can be represented by richer features — current 8h/512d may not have enough capacity to fill 256 slice tokens informatively.
- Sweep slice count below 128: try `--model-slices 64` and `96`. The negative result here suggests "fewer, sharper slices" might do better than "more, dilute slices" on this 4 L backbone.
- If 256 slices is revisited, also raise `--lr-warmup-epochs` to 1–2: the curve peaked at epoch 7 (≈25 % of step budget), suggesting the model overfit/plateaued before the cosine schedule could fully decay; warmup might delay the early peak.
- Architectural follow-up: replace the slice-attention K-set (which we just doubled) with an *adaptive* slice count (mean field of slices selected per-batch) so the model picks how granular to be, instead of being forced to 256.

### Bug-flag for advisor
The PR command used `--cosine-t-max 50` but the valid argparse flag is `--lr-cosine-t-max`. With the original, argparse aborts with `error: unrecognized arguments` (saw this in the first launch attempt log `round10-slices-256_20260430_233316.log`). I corrected to `--lr-cosine-t-max 50` for the recorded run. Worth fixing in PR templates so future students don't trip on it.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #139 because ADVISOR: Closing this slices=256 experiment. Test 12.389 vs PR #115 SOTA 10.580 = **+17.1% worse**. The hypothesis (more slices → finer high-frequency capture for tau_y/z) failed cleanly — every component regressed by 4-15%, with no selective improvement on the targeted tau axes.

**Important context:** When you launched, baseline was PR #111 (11.142). PR #115 (compound lr=1e-4 + EMA=0.999) merged during your run and raised the bar by −5.04%. You compared against PR #111 (+11.2%); vs current SOTA is +17.1%.

**Why slices=256 lost:**
1. **Memory pressure (94.4 GB / 97 GB peak, 98.8% allocated)** — the model was at the edge of OOM, suggesting attention overhead grew quadratically with slice count. This left no margin for batch_size or other capacity scaling.
2. **Per-epoch slowdown ~2×** (et=8.9 min vs SOTA 30 min) — wait, looking again: total runtime 270m, ~30 epochs… that's actually faster per-epoch for this config. Looking more carefully: best val at ep7 = 11.21 (vs 9.99 at ep9 for 128-slice SOTA). Even at faster epoch cadence, the model converged to a *worse* asymptote.
3. **Slice attention quadratic in slice count** — 256 slices means each slice attends to 256 others. The information bottleneck per slice may have actually decreased (each slice now represents a smaller geometric subset), making it harder to learn long-range structure.
4. **Per-axis hypothesis failed selectively too.** tau_y +12.7%, tau_z +12.3% — equally bad as surface metrics. Slice density did not specifically help high-frequency components.

**Note on your config:** Your run used lr=5e-5 + vol_w=2.0 + EMA=0.999 + T_max=50, NOT the PR #115 SOTA stack (lr=1e-4 + EMA=0.999 + vol_w=1.0). So the slices delta is convolved with the lr/vol_w differences. The slice lever still failed clearly enough that it's not worth a strict-SOTA-stack retry.

**Conclusion:** model_slices=128 is the right setting. Lever closed.

**Fern reassigned: Lion β1=0.95** — untested optimizer hyperparameter, single CLI delta from PR #115 SOTA..
